lib/src/value/color.ts

```diff
@@ -3,202 +3,372 @@
+      options.alpha === null
+        ? NaN
+        : options.alpha === undefined
         ? 1
-        : fuzzyAssertInRange(color.alpha, 0, 1, 'alpha');
+        : fuzzyAssertInRange(options.alpha, 0, 1, 'alpha');
+
+    switch (space) {
+      case 'srgb':
+        if (this.format === 'rgb') {
+          // Convert from 0-255 to 0-1
+          this.color = getColor({
+            spaceId: space,
+            coords: [
+              (options.red ?? NaN) / 255,
+              (options.green ?? NaN) / 255,
+              (options.blue ?? NaN) / 255,
+            ],
+            alpha,
+          });
+        } else {
+          this.color = getColor({
+            spaceId: space,
+            coords: [
+              options.red ?? NaN,
+              options.green ?? NaN,
+              options.blue ?? NaN,
+            ],
+            alpha,
+          });
+        }
+        break;
+      case 'srgb-linear':
+      case 'display-p3':
+      case 'a98-rgb':
+      case 'prophoto-rgb':
+      case 'rec2020':
+        this.color = getColor({
+          spaceId: space,
+          coords: [
+            options.red ?? NaN,
+            options.green ?? NaN,
+            options.blue ?? NaN,
+          ],
+          alpha,
+        });
+        break;
+
+      case 'hsl':
+        this.color = getColor({
+          spaceId: space,
+          coords: [
+            options.hue ?? NaN,
+            options.saturation ?? NaN,
+            options.lightness ?? NaN,
+          ],
+          alpha,
+        });
+        break;
+
+      case 'hwb':
+        this.color = getColor({
+          spaceId: space,
+          coords: [
+            options.hue ?? NaN,
+            options.whiteness ?? NaN,
+            options.blackness ?? NaN,
+          ],
+          alpha,
+        });
+        break;
+
+      case 'lab':
+      case 'oklab':
+        this.color = getColor({
+          spaceId: space,
+          coords: [
+            options.lightness ?? NaN,
+            options.a ?? NaN,
+            options.b ?? NaN,
+          ],
+          alpha,
+        });
+        break;
+
+      case 'lch':
+      case 'oklch':
+        this.color = getColor({
+          spaceId: space,
+          coords: [
+            options.lightness ?? NaN,
+            options.chroma ?? NaN,
+            options.hue ?? NaN,
+          ],
+          alpha,
+        });
+        break;
+
+      case 'xyz':
+      case 'xyz-d65':
+      case 'xyz-d50':
+        this.color = getColor({
+          spaceId: space,
+          coords: [options.x ?? NaN, options.y ?? NaN, options.z ?? NaN],
+          alpha,
+        });
+        break;
+    }
   }
 
   /** `this`'s red channel. */
   get red(): number {
-    if (this.redInternal === undefined) {
-      this.hslToRgb();
+    emitColor4ApiDeprecation('red');
+    try {
+      return get(this.color, 'red');
+    } catch (error) {
+      return get(to(this.color, 'srgb'), 'red');
     }
-    return this.redInternal!;
   }
```

Comment:
Currently, these will return the channel as-is for non-legacy RGB colors, instead of being converted to legacy RGB and returning that value. That's probably _not_ desirable—legacy code that's using these getters is generally going to be code that's assuming everything is convertible to legacy RGB anyway. Similarly, converting to sRGB first won't solve this completely, because sRGB's channels are in [0, 1] but RGB's are in [0, 255]

The spec is vague on this point, but it should probably update the definitions for the legacy channel getters to explicitly mention that they convert colors to legacy spaces first.

Reply:
Spec adjusted in https://github.com/sass/sass/pull/3736, embedded host adjusted in https://github.com/sass/embedded-host-node/pull/259/files/3713dee14d58ffd14f6ce574b9a7c5bfc8504bbf..04010b6960e5c45fc3f5859655a79f91da3efc4c